### PR TITLE
JIT: Clean up gtCloneExpr

### PR DESCRIPTION
**Pull request by jakobbotsch**
- Remove its ability to replace locals by constants while cloning. Only loop unrolling uses this capability and it can be replaced by a simple visit over the tree after cloning, so that everyone else does not need to pay for it.
- Remove its ability to add flags. This is only used by hoisting to add `GTF_MAKE_CSE` flag recursively to the cloned tree. I do not see any good reason not just to only put this on the root node.
- Simplify how it propagates flags; simply copy them from the source node instead of having a separate `gtUpdateNodeSideEffects` call for the cloned tree. I do not see why the cloned version should have its flags updated like this when the original shouldn't.

A few diffs expected due to not marking hoisted clones with `GTF_MAKE_CSE` recursively, which sometimes allows us to optimize them away. Some other diffs from the removal of `gtUpdateNodeSideEffects(copy)` in `gtCloneExpr`, which can impact cases where we get our flag maintenance wrong (e.g. #13758).

### CODE REVIEW

**Comment by jakobbotsch** [comment]
cc @dotnet/jit-contrib PTAL @EgorBo, maybe @AndyAyersMS too since you've been looking at `GTF_MAKE_CSE` recently.

Should be ready pending CI.

[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=538808&view=ms.vss-build-web.run-extensions-tab). Some from the `GTF_MAKE_CSE` change, which allows some nested trees to be removed by assertion prop now (previously `gtNodeHasSideEffects` would return true for them). Some other diffs because of the removal of the removal of the `gtUpdateNodeSideEffects` call, as mentioned above.

File: src/coreclr/jit/optimizer.cpp [hunk 1]
```diff
@@ -1798,6 +1800,70 @@ bool Compiler::optTryUnrollLoop(FlowGraphNaturalLoop* loop, bool* changedIR)
...
+// Arguments:
+//   block   - The block to replace in
+//   lclNum  - The local to replace
+//   cnsVal  - The constant to replace with
+//
+// Remarks:
+//   This is used to replace the loop iterator with the constant value when
+//   unrolling.
+//
+void Compiler::optReplaceScalarUsesWithConst(BasicBlock* block, unsigned lclNum, ssize_t cnsVal)
```
Comment (EgorBo) [inline src/coreclr/jit/optimizer.cpp:1816]: I am not familair with the existing logic - can you explain why it's fine to do propagation like this? can the locals be address exposed?

**Review by EgorBo — APPROVED** [review]
LGTM otherwise

**Comment by AndyAyersMS** [comment]
GTF_MAKE_CSE change is good, I had the same thing in the branch I was using to investigate why we skip doing some of them.

File: src/coreclr/jit/optimizer.cpp (hunk 1 above)
Comment (jakobbotsch) [reply src/coreclr/jit/optimizer.cpp:1816]: `FlowGraphNaturalLoop::AnalyzeIteration` checks for this (it guarantees that the iterator local is only defined once within the loop and that the loop invariant is always true).

**Comment by jakobbotsch** [comment]
Failures are #97437

**Pull request merged**